src/types.ts

```diff
@@ -4290,6 +4290,395 @@ export class ZodPipeline<
+
+    return new ZodTemplateLiteral({
+      ...this._def,
+      parts,
+      regexString: this._appendToRegexString(this._def.regexString, part),
+    });
+  }
+
+  protected _appendToRegexString(
+    regexString: string,
+    part: TemplateLiteralPrimitive | TemplateLiteralInterpolatedPosition
+  ): string {
+    return `^${this._unwrapRegexString(
+      regexString
+    )}${this._transformPartToRegexString(part)}$`;
+  }
+
+  protected _transformPartToRegexString(
+    part: TemplateLiteralPrimitive | TemplateLiteralInterpolatedPosition
+  ): string {
+    if (!(part instanceof ZodType)) {
+      return this._getEscapedStringForRegex(part);
+    }
+
+    if (part instanceof ZodLiteral) {
+      return this._getEscapedStringForRegex(part._def.value);
+    }
+
+    if (part instanceof ZodString) {
+      return this._transformZodStringPartToRegexString(part);
+    }
+
+    if (part instanceof ZodEnum || part instanceof ZodNativeEnum) {
+      const values =
+        part instanceof ZodEnum
+          ? part._def.values
+          : util.getValidEnumValues(part._def.values);
+
+      return `(${values.map(this._getEscapedStringForRegex).join("|")})`;
+    }
+
+    if (part instanceof ZodUnion) {
+      return `(${(part._def.options as any[])
+        .map((option) => this._transformPartToRegexString(option))
+        .join("|")})`;
+    }
+
+    if (part instanceof ZodNumber) {
+      return this._transformZodNumberPartToRegexString(part);
+    }
+
+    if (part instanceof ZodOptional) {
+      return `(${this._transformPartToRegexString(part.unwrap())})?`;
+    }
+
+    if (part instanceof ZodTemplateLiteral) {
+      return this._unwrapRegexString(part._def.regexString);
+    }
+
+    if (part instanceof ZodBigInt) {
+      // FIXME: include/exclude '-' based on min/max values after https://github.com/colinhacks/zod/pull/1711 is merged.
+      return "\\-?\\d+";
+    }
+
+    if (part instanceof ZodBoolean) {
+      return "(true|false)";
+    }
+
+    if (part instanceof ZodNullable) {
+      do {
+        part = part.unwrap();
+      } while (part instanceof ZodNullable);
+
+      return `(${this._transformPartToRegexString(part)}|null)${
+        part instanceof ZodOptional ? "?" : ""
+      }`;
+    }
+
+    if (part instanceof ZodBranded) {
+      return this._transformPartToRegexString(part.unwrap());
+    }
+
+    if (part instanceof ZodAny) {
+      return ".*";
+    }
+
+    if (part instanceof ZodNull) {
+      return "null";
+    }
+
+    if (part instanceof ZodUndefined) {
+      return "undefined";
+    }
+
+    throw new Error(
+      "Zod error: Unsupported ZodTemplateLiteral argument! Please submit an issue at https://github.com/colinhacks/zod/issues"
+    );
+  }
+
+  protected _transformZodStringPartToRegexString(part: ZodString): string {
+    let maxLength = Infinity,
+      minLength = 0,
+      endsWith = "",
+      startsWith = "";
+
+    for (const ch of part._def.checks) {
+      if (ch.kind === "cuid") {
```

Comment:
> I think here we may be not handling the trim() ZodStringCheck path.

Haven't thought about this use case. 🤔
The general rule is, any transforms are prohibited (see `ZodEffects`/`ZodPipeline` ban), let's go with that.. and see if consumers are interested in a different behavior here?

> I found it while checking if this could be rewritten as an exhaustive switch() check, which I suggest we do to find precisely these specific scenarios!

I tried to steer away from this, because it'll end up hurting every new feature. It'll make contributing and reviewing new features in other parts of the codebase harder, because now a lot of PRs will have to deal with template literal support - painful context switch.

I find that slowly adding support for new things is also safer for this feature. New things will get more focused PRs with attention to properly testing standalone cases, and mixed cases (different types/checks).

WDYT?

Reply:
> The general rule is, any transforms are prohibited (see ZodEffects/ZodPipeline ban), let's go with that

sounds good to me!

> I tried to steer away from this, because it'll end up hurting every new feature.
> We could throw on unsupported checks.. without forcing to implement every check

Hmm. In one hand, _because_ of the exhaustive checks I found the `trim()` scenario. But on the other hand, I do see what you mean in terms of making each feature more complicated, especially with all the definitions being in one big file... :) 

The catch-all you added at the end is a good middle ground. Thank you!